bundles/net.tourbook/src/net/tourbook/ui/views/rawData/DialogReimportTours.java

```diff
@@ -60,57 +72,64 @@
 import org.eclipse.swt.widgets.DateTime;
 import org.eclipse.swt.widgets.Display;
 import org.eclipse.swt.widgets.Group;
+import org.eclipse.swt.widgets.Label;
 import org.eclipse.swt.widgets.Shell;
 
 public class DialogReimportTours extends TitleAreaDialog {
```

Comment:
When reimporting  one selected tour with these settings...

![image](https://user-images.githubusercontent.com/1283445/126028051-a6c3d2f4-21b3-4208-bb6e-aa7e7f1a46b6.png)

...this exception occures -> I've not further tested the reimporting

```

java.lang.NullPointerException
	at net.tourbook.ui.views.rawData.DialogReimportTours.actionReimportSelectedTours(DialogReimportTours.java:306)
	at net.tourbook.ui.views.rawData.DialogReimportTours.doReimport(DialogReimportTours.java:935)
	at net.tourbook.ui.views.rawData.DialogReimportTours.lambda$9(DialogReimportTours.java:1096)
	at net.tourbook.ui.views.rawData.DialogReimportTours$$Lambda$723/0x0000000000000000.run(Unknown Source)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:74)
	at net.tourbook.ui.views.rawData.DialogReimportTours.okPressed(DialogReimportTours.java:1063)
	at org.eclipse.jface.dialogs.Dialog.buttonPressed(Dialog.java:468)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.jface.dialogs.Dialog$$Lambda$124/0x0000000000000000.accept(Unknown Source)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4209)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1043)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4026)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3626)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at net.tourbook.ui.views.rawData.ActionReimportTours.run(ActionReimportTours.java:44)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.jface.action.ActionContributionItem$$Lambda$377/0x0000000000000000.handleEvent(Unknown Source)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4209)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1043)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4026)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3626)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1157)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.ui.internal.Workbench$$Lambda$122/0x0000000000000000.run(Unknown Source)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at net.tourbook.application.Application.start(Application.java:35)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:654)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:591)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1462)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1435)

```

Reply:
why do you merge my PR if there are still changes to do ?

Reply:
Can you try the concurrent import ?

Reply:
> 
> 
> why do you merge my PR if there are still changes to do ?

Can you mark PR's as draft or something else when I should not review them

Reply:
> 
> 
> Can you try the concurrent import ? (Click on the bottom checkbox "Skip tours for which the reimport file is not found)

I'll try it tomorrow

Reply:
Found these issues

1. Must be run concurrent, this is an expensive action
2. `_coundDownLath` is not updated -> endless loop, I also saw the same issue in the concurrent method

![image](https://user-images.githubusercontent.com/1283445/126061230-f1a97107-ea70-438a-9b98-48ca6b4f379a.png)

The tour reimport feature is not yet suited to be fully concurrent, this database lock should not be needed. 

So a bigger rework of the import/reimport feature is necessary to speed also the tour saving method with all it's sub methods, e.g. update cached tour titles

![image](https://user-images.githubusercontent.com/1283445/126061384-66404d6d-03d2-4914-a85d-95152a2616f8.png)

Reply:
>The tour reimport feature is not yet suited to be fully concurrent, this database lock should not be needed.

Ok, I will remove it.

>So a bigger rework of the import/reimport feature is necessary to speed also the tour saving method with all it's sub methods, e.g. update cached tour titles

I don't understand fully. Can you give more details please ?

Reply:
> 
> 
> When reimporting one selected tour with these settings...
> 
> ![image](https://user-images.githubusercontent.com/1283445/126028051-a6c3d2f4-21b3-4208-bb6e-aa7e7f1a46b6.png)
> 
> ...this exception occures -> I've not further tested the reimporting
> 
> ```
> 
> java.lang.NullPointerException
> 	at net.tourbook.ui.views.rawData.DialogReimportTours.actionReimportSelectedTours(DialogReimportTours.java:306)

Fixed and will be available in next PR

Reply:
> > So a bigger rework of the import/reimport feature is necessary to speed also the tour saving method with all it's sub methods, e.g. update cached tour titles
> 
> I don't understand fully. Can you give more details please ?

The following method is called when a tour is saved

https://github.com/wolfgang-ch/mytourbook/blob/b59ca259e0a411ee510cf36424818394f02769ca/bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java#L2637

which contains these methods

https://github.com/wolfgang-ch/mytourbook/blob/b59ca259e0a411ee510cf36424818394f02769ca/bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java#L2747-L2749

and ` updateCachedFields` is not concurrent

https://github.com/wolfgang-ch/mytourbook/blob/b59ca259e0a411ee510cf36424818394f02769ca/bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java#L2892-L2900

So removing the database lock would need a lot of adjustments. 

My approach would be to implement a new concurrent tour save method otherwise you have to verify all existing tour save possibilities. 

The new tour save method should first save all tours concurrently and then do the post methods, e.g. ` updateCachedFields`

Reply:
Ok, thanks for the insight, I will start looking at this.

Reply:
While I was not able to exactly implement it as described, you can see my tour reimport rework [here](https://github.com/wolfgang-ch/mytourbook/pull/377)